src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,513 @@
+        }
+
+        private Entry[] Resize()
+        {
+            int count = Count;
+            int newSize = HashHelpers.ExpandPrime(count);
+            var entries = new Entry[newSize];
+            Array.Copy(_entries, 0, entries, 0, count);
+
+            var newBuckets = new int[newSize];
+            while (count-- > 0)
+            {
+                int bucketIndex = (int)(GetHashCode(entries[count].key) % (uint)newBuckets.Length);
+                entries[count].next = newBuckets[bucketIndex] - 1;
+                newBuckets[bucketIndex] = count + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = entries;
+
+            return entries;
+        }
+
+        public KeyEnumerable Keys => new KeyEnumerable(this);
+
+        public ValueEnumerable Values => new ValueEnumerable(this);
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        {
+            Entry[] entries = _entries;
+            int i = 0;
+            int count = Count;
+            while (count > 0)
+            {
+                if (entries[i].next > -2)
+                {
+                    count--;
+                    array[index++] = new KeyValuePair<TKey, TValue>(
+                        entries[i].key,
+                        entries[i].value);
+                }
+                i++;
+            }
+        }
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() => new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index, _count;
+
+            internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _index = 0;
+                _count = _dictionary.Count;
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_count == 0)
+                {
+                    Current = default;
+                    return false;
+                }
+
+                _count--;
+
+                while (_dictionary._entries[_index].next < -1)
+                    _index++;
+
+                Current = new KeyValuePair<TKey, TValue>(
+                    _dictionary._entries[_index].key,
+                    _dictionary._entries[_index++].value);
+                return true;
+            }
+
+            public KeyValuePair<TKey, TValue> Current { get; private set; }
+            object IEnumerator.Current => Current;
+            void IEnumerator.Reset()
+            {
+                _index = 0;
+                _count = _dictionary.Count;
+            }
+            public void Dispose() { }
+        }
+
+        public struct KeyEnumerable : IEnumerable<TKey>
```

Comment:
Rename to `KeyCollection` and join with `KeyEnumerator`?

Reply:
I think it's fine for `Keys` and `Values` to not implement any of the collectoin interfaces. Interfaces that mostly throw are bad patterns. The only useable things they add are Count and Contains (which is on dictionary already as Count and ContainsKey) and CopyTo (which I have never needed to use anyway is trivially implemented). We are missing an equivalent of `Values.Contains` but `ContainsValue` is a perf antipattern anyway.

Reply:
Note if we ever want DictionarySlim to implement IDictionary, we would at that time be forced to implement `ICollection<TKey> Keys` and `ICollection<TValue> Values`.

Reply:
I'm tempted to add ToArray(). It's often used and can be much faster from here than IEnumerable.

Reply:
Ah - there's the reason to implement ICollection with at least Count functional. So that ToArray on IEnumerable can pre allocate. It casts to ICollection if it can:
https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/ToCollection.cs#L20
That would be better than special casing ToArray here.